sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java

```diff
@@ -506,30 +506,55 @@ void replaceOutputs(Map<PValue, ReplacementOutput> originalToReplacement) {
      * <p>Provides an ordered visit of the input values, the primitive transform (or child nodes for
      * composite transforms), then the output values.
      */
-    private void visit(PipelineVisitor visitor, Set<PValue> visitedValues) {
+    private void visit(
+        PipelineVisitor visitor,
+        Set<PValue> visitedValues,
+        Set<Node> visitedNodes,
+        Set<Node> passedComposites) {
+      if (getEnclosingNode() != null && !visitedNodes.contains(getEnclosingNode())) {
```

Comment:
This is the "Always have entered an enclosing node before visiting any enclosed nodes" invariant